### DISCUSSION TITLE
Using JupyterLite components

### DISCUSSION THREAD

**Discussion by sswatson** [jupyterlite/jupyterlite#1050] [Q&A] [answered]
I have an application with a custom notebook environment, and I'd like to add a Python cell type using some of the components of JupyterLite. This means that I would need to use JupyterLab's `OutputArea` and set up communication between the kernel and the output area.

I was able to do this with regular Jupyter along the lines of [Juniper](https://github.com/ines/juniper), but I'm struggling to figure out how to replace the remote kernel with a Pyodide-based kernel.

Has anyone tried to do this, or does anyone have pointers for where I should look to understand better what I might need to do?

**Comment by bollwyvl** [comment, marked as answer]
At present, it is not possible to mix browser kernels and server kernels, as the [`ServiceManager`](https://github.com/jupyterlite/jupyterlite/blob/8ebd346a2d5550d7f8a224692db818dfd34cb1d9/packages/server/src/app.ts#L25) that controls this behavior is winner-take-all, and the jupyterlite "server" needs to wrest total control for just about everything (contents, kernspec listing, kernel "websockets", etc.) The [upstream issue](https://github.com/jupyterlab/jupyterlab/issues/13631) (with linked PR) demonstrates some of these complexities.

The current lowest-effort offering would be to use an embedded iframe of a customized jupyterlite app, starting from the simplest example (perhaps the `repl`, though it can't do something, like widgets). Alternately, a more full application (like `retro`) can also be hacked down with CSS to achieve a closer experience. The shortcoming here is that _each_ iframe would be its own application, with its own kerenls, as there is not yet a way to host _another_ iframe that just provides the "server" piece, though it would be _possible_ to do so, as all of the boundaries between the "client" and "server" move over standards-based protocols (`fetch`, `postMessage`, websockets, WebWorker).

Providing directly-embedded application components has been out of scope, as the focus of this effort is to provide an end user experience that most closely resembles JupyterLab, partially so that [many existing prebuilt lab extensions](https://pypi.org/search/?q=&o=&c=Framework+%3A%3A+Jupyter+%3A%3A+JupyterLab+%3A%3A+3&c=Framework+%3A%3A+Jupyter+%3A%3A+JupyterLab+%3A%3A+Extensions+%3A%3A+Prebuilt) work without modification... but as all the jupyterlab and jupyterlite components are published on npm, there are likely more examples out there that I just don't know about.

**Reply by sswatson** [reply]
Thanks for the reply; that's helpful to know. I can tell that the focus has been at the application level rather than the component level, which makes sense.

Fortunately I'm not looking to mix browser kernels and server kernels, as the main problem is that the server kernel is extremely slow and I just want to replace it with a browser one. Also, my component just needs to be able to send code to the kernel, get the result object back, and hand it to an OutputArea. All of the other notebook functionality is managed by the host app. So I'm still hopeful that there's a solution for this use case.

If I end up coming up with something usable and shareable, I will make sure to post a link on this thread in case it's helpful to someone else.
